File: leetcode/328-odd-even-linked-list/odd-even-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: ListNode) -> ListNode:
        oddHead = head

        if head != None:
            evenHead = head.next

            odd = oddHead
            even = evenHead

            while even != None and even.next != None:
                odd.next = odd.next.next
                odd = odd.next
                even.next = even.next.next
                even = even.next

            odd.next = evenHead

        return oddHead

File: leetcode/328-odd-even-linked-list/odd-even-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
const oddEvenList = function (head) {
  const oddHead = head;

  if (!head) {
    return null;
  }

  const evenHead = head.next;

  let oddP = oddHead;
  let evenP = evenHead;

  while (oddP.next && evenP.next) {
    const oddNext = oddP.next.next;
    const evenNext = evenP.next.next;
    oddP.next = oddNext;
    evenP.next = evenNext;
    oddP = oddP.next;
    evenP = evenP.next;
  }

  oddP.next = evenHead;

  return oddHead;
};
